app/app.go

```diff
@@ -423,15 +423,25 @@ func New(
 
 	baseAppOptions = memiavlstore.SetupMemIAVL(logger, homePath, appOpts, false, false, baseAppOptions)
 
-	blockProposalHandler := NewProposalHandler(encodingConfig.TxConfig.TxDecoder(), identity)
+	txSelector := NewExtTxSelector(
+		baseapp.NewDefaultTxSelector(),
+		encodingConfig.TxConfig.TxDecoder(),
+		identity)
 
 	// NOTE we use custom transaction decoder that supports the sdk.Tx interface instead of sdk.StdTx
 	// Setup Mempool and Proposal Handlers
 	baseAppOptions = append(baseAppOptions, func(app *baseapp.BaseApp) {
 		mempool := mempool.NoOpMempool{}
 		app.SetMempool(mempool)
-		app.SetPrepareProposal(blockProposalHandler.PrepareProposalHandler())
-		app.SetProcessProposal(blockProposalHandler.ProcessProposalHandler())
+
+		// Re-use the default prepare proposal handler, extend the transaction validation logic
+		defaultProposalHandler := baseapp.NewDefaultProposalHandler(mempool, app)
+		defaultProposalHandler.SetTxSelector(txSelector)
+		app.SetPrepareProposal(defaultProposalHandler.PrepareProposalHandler())
+
+		// The default process proposal handler do nothing when the mempool is noop,
+		// so we just implement a new one.
+		app.SetProcessProposal(defaultProposalHandler.ProcessProposalHandler())
```

Comment:
The custom process proposal logic is critical.

Reply:
The critical logic is moved inside ExtTxSelector, no more custom process proposal handler is needed.